python-lib/ontology_tagging/formatter_by_document.py

```diff
@@ -0,0 +1,442 @@
+        else:
+            self.tag_keywords.append(np.nan)
+        self.tag_sentences.append(" ".join(unique_list(matched_sentence)))
+        self.output_df = (
+            self.output_df.append(tag_columns_for_json, ignore_index=True)
+            if one_row_per_doc_json
+            else self.output_df.append(
+                line,
+                ignore_index=True,
+            )
+        )
+
+    def _get_tags_in_row_category(
+        self, match: Span, line: dict, line_full: dict, sentence: Doc, language: AnyStr
+    ) -> Tuple[dict, dict]:
+        """
+        Called by _write_row_category
+        Enrich 'line' and 'line_full' dictionaries with new found tags
+
+        Args:
+            line (dict): dictionary of category (keys) and tags (values)
+            line_full (dict): dictionary of category (keys).
+                Each category has for value a dictionary of tags and their associated datas (keywords, sentences, occurences of the tag)
+            sentence (Doc): sentence that matched with keyword(s)
+            language (AnyStr): language of the sentence
+
+        Returns:
+            - Tuple(dict,dict) : The enriched dictionaries 'line' and 'line_full'
+
+        """
+        keyword = match.text
+        tag = match.ent_id_
+        category = match.label_
+        sentence = sentence
+        if tag not in line_full[category]:
+            line_full[category][tag] = {
+                "count": 1,
+                "sentences": [sentence],
+                "keywords": [keyword],
+            }
+
+            line[category].append(tag)
+        else:
+            line_full[category][tag]["count"] += 1
+            if sentence not in line_full[category][tag]["sentences"]:
+                line_full[category][tag]["sentences"].append(sentence)
+            if keyword not in line_full[category][tag]["keywords"]:
+                line_full[category][tag]["keywords"].append(keyword)
+        return line, line_full
+
+    def _merge_df_columns_category(
+        self, input_df: pd.DataFrame, text_column: AnyStr
+    ) -> pd.DataFrame:
+        """
+        Called by _write_row_category, when the format is 'one row per document (array columns)' and there are categories.
+        Insert columns 'tag_keywords' and 'tag_sentences' into self.output_df
+
+        Returns:
+            pd.DataFrame: the complete output dataframe after setting the columns in the right order
+
+        """
+        tag_list_columns = self.output_df.columns.tolist()
+        tag_list_columns_unique = generate_unique_columns(
+            df=self.output_df,
+            columns=[
+                unicode_normalize_text(text=column) for column in tag_list_columns
+            ],
+            prefix="tag_list",
+        )
+        self.output_df.columns = tag_list_columns_unique
+        for column, column_unique in zip(tag_list_columns, tag_list_columns_unique):
+            category = column.split("_")[-1]
```

Comment:
What if the category contains a "_" , this seems a bit flaky to get the category from the column name

Reply:
According to our earlier discussion,I've fiwed it  here b9c1ad3